design/one-pager-function-go-templating.md

```diff
@@ -0,0 +1,171 @@
+* Status: Accepted
+
+## Background
+
+With Crossplane 1.14, Composition Functions will be promoted to `v1beta1`, and several 
+composition functions will be introduced as an extension point to be used in pipelines. 
+One of these extensions is the [Go Templating Function], which allows users to render Crossplane
+resources using Go templating capabilities like conditionals and loops. In addition to that, 
+users can use values in the environment configs or other resource fields to create/configure
+Crossplane resources.
+
+## Getting Template Files
+
+With the initial implementation of the Go Templating Function, users will be able
+to provide templates in two different ways:
+- An `inline` source can be used to provide template files as a string in the
+  `input.inline.template` field of the pipeline step.
+- A `fileSystem` source can be used to provide template files with folder path
+  in the `input.fileSystem.dirPath` field of the pipeline step.
+
+Both of these options have their own advantages and disadvantages. While `inline` source
+is faster and easier to use, it is not suitable for large template files and hard to maintain.
+On the other hand, `fileSystem` source requires the files to be present in the local file system,
+which might be possible by mounting config maps with `DeploymentRuntimeConfig` or building another
+container image on top of the existing one with the template files. 
+
+In the next iterations, it is planned to support fetching template files from a remote location like
+GitHub repositories or S3 buckets.
+
+## Rendering Templates and Accessing Resources
+
+To make rendering templates and accessing resource fields easier, function requests
+will be converted to `map[string]interface{}` and passed to the template engine. This will
+allow users to access resource fields in the template files like:
+
+```yaml
+{{- range $i := until ( .observed.composite.resource.spec.count | int ) }}
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: User
+metadata:
+  annotations:
+    gotemplating.fn.crossplane.io/composition-resource-name: test-user-{{ $i }}
+  name: test-user-{{ $i }}
+  labels:
+    testing.upbound.io/example-name: test-user-{{ $i }}
+spec:
+  forProvider: {}
+
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: AccessKey
+metadata:
+  name: sample-access-key-{{ $i }}
+spec:
+  forProvider:
+    userSelector:
+      matchLabels:
+        testing.upbound.io/example-name: test-user-{{ $i }}
+  writeConnectionSecretToRef:
+    name: sample-access-key-secret-{{ $i }}
+    namespace: crossplane-system
+{{- end }}
+
+```
+
+To compare and prepare the desired composed resources map with consistent keys, `function-go-templating` requires a special
+annotation `gotemplating.fn.crossplane.io/composition-resource-name` to be assigned to the input resources. This
+annotation will match the resource name in the template file with the resource name in the desired resources.
+
+## Configuring Composite Connection Details
```

Comment:
Can you help me understand the motivation behind this design? Is the providerConfigRef(Name) still being used to connect to the provider and this serves different purpose? Also what is `meta resource` :)?

Just to make sure the terminology is the same, in the platform team we would typically use the term `composite` to mean the actual composition and separately XR to mean the `composite resource`. In this case the `CompositeConnectionDetails` provides connection details for the XR or the composition?

Reply:
> Can you help me understand the motivation behind this design? Is the providerConfigRef(Name) still being used to connect to the provider and this serves different purpose?

Sorry, I couldn't understand this question. Can you explain a bit more?

> Just to make sure the terminology is the same, in the platform team we would typically use the term composite to mean the actual composition and separately XR to mean the composite resource. In this case the CompositeConnectionDetails provides connection details for the XR or the composition?

`CompositeConnectionDetails` is a meta resource which configures composite resource's connection details. And by meta resource what I mean is, it's not a real k8s resource. But it contains metadata information for other resources.

Reply:
Thank you for clarifying. What I meant by my first question is that I'm trying to understand what is the purpose of the  `CompositeConnectionDetails` and why do we need it. Is it somehow equivalent to `[writeConnectionSecretsToNamespace](https://docs.crossplane.io/v1.13/concepts/compositions/#hl-15-5)`? Most probably I'm just misunderstanding something.

Reply:
It's equivalent to [here](https://docs.crossplane.io/v1.13/concepts/compositions/#hl-18-14). Assume that you define several resources in your composition and want to propagate some of their connection details like username/password from one resource and endpoint from another. In that case, you can use this meta resource to collect different information from different resources and write it to composite resource's connection secret (the one created with the UID of the composite resource).